maxTotalRows should be checked in DataSourceCompactionConfig before setting targetCompactionSizeBytes

jihoonson opened the pull request:
Fixes https://github.com/apache/incubator-druid/issues/7356.

In auto compaction, `maxTotalRows` and `maxRowsPerSegment` are not supposed to set if `targetCompactionSizeBytes` is set. But, if `targetCompactionSizeBytes` and `maxTotalRows` are set to null, both are set to some default values which makes submitting compaction tasks failed. In this PR, I fixed this bug and added integration tests for auto compaction.

surekhasaharan commented on server/src/main/java/org/apache/druid/client/indexing/ClientCompactQueryTuningConfig.java:
```diff
@@ -31,15 +31,15 @@
 {
   // These default values should be synchronized with those of IndexTuningConfig
   private static final int DEFAULT_MAX_ROWS_IN_MEMORY = 75_000;
```
This value got changed to 1_000_000, may should get this from [here](https://github.com/apache/incubator-druid/blob/master/server/src/main/java/org/apache/druid/segment/indexing/TuningConfig.java#L37). Unless you need it to be 75_000.

jihoonson replied:
Thanks. I found this is not necessary because they will be set to defaults in `IndexTuningConfig` if they are null. Removed all default values.

surekhasaharan commented on server/src/main/java/org/apache/druid/server/coordinator/DataSourceCompactionConfig.java:
```diff
@@ -104,6 +100,43 @@ public DataSourceCompactionConfig(
+  /**
+   * Copied from CompactionTask#getValidTargetCompactionSizeBytes.
+   * Must be synced.
+   */
```
May be not in this PR but better to create a common method for both ?

jihoonson replied:
Hmm, they are not exactly same because `numShards` is missing in `DataSourceCompactionConfig`. I'm not sure how to do it well.

Rather, I think we need to merge `server` and `indexing-service` modules somehow, so that the coordinator and the overlord use the same classes for ingestion spec. The biggest problem here is the dependency on `druid-indexing-hadoop` which isn't necessary for all modules depending on `server`.

surekhasaharan commented on integration-tests/src/test/java/org/apache/druid/tests/indexer/AbstractIndexerTest.java:
```diff
@@ -112,7 +112,7 @@ protected void waitForAllTasksToComplete()
   }
 
-  protected String getTaskAsString(String file) throws IOException
+  protected String getResourceAsString(String file) throws IOException
```
Nice catch !

jihoonson commented:
I removed new integration tests I added in this PR because they are flaky and I think it would take some time to fix them. I don't think it's important to block the release. I'll raise a follow-up PR for integration tests.

jon-wei commented on server/src/main/java/org/apache/druid/server/coordinator/DataSourceCompactionConfig.java:
```diff
@@ -104,6 +100,43 @@ public DataSourceCompactionConfig(
+
+  /**
+   * Copied from CompactionTask#hasPartitionConfig.
+   * Must be synced.
```
Can you add a comment here and elsewhere explaining why it's copied and has to be synced?

jihoonson replied:
Thanks, added.

jihoonson commented:
I ran unit tests and integration tests locally and they passed. I also tested the bug has fixed in my local cluster.

jon-wei commented:
TravisCI is still having issues today, merging as TeamCity has passed and local UT/IT have passed.
